File: typescript-examples/src/policy-management-examples.ts
import { PolicyDataListResponse, PolicyListResponse, PolicyManagementApi, ServerConfiguration, createConfiguration } from "@privatedataservices/metronome";

const clientConfig = createConfiguration( {
    baseServer: new ServerConfiguration("pds-internal-code-examples", "eu-north-1", {}, true),
    authMethods: { ApiKeyAuth: process.env.METRONOME_API_KEY }
});
const pmApiClient = new PolicyManagementApi(clientConfig);

const createPolicyExample = async () => {
    try {
        const response = await pmApiClient.putPolicy("user-account-access", {"agingStrategy": "since-last-read-or-write", "agingOffsetAmount": 2, "agingOffsetUnit": "year"});
        console.log("Policy %s created successfully.  Fields updated were: %s", response.policyId, JSON.stringify(response.fieldsUpdated));
    } catch (err) {
        console.error("Failed to create new policy because: %s", JSON.stringify(err));
    }
}

const modifyPolicyExample = async () => {
    try {
        const response = await pmApiClient.putPolicy("user-account-access", {"agingStrategy": "since-last-read-or-write", "agingOffsetAmount": 3, "agingOffsetUnit": "month"});
        console.log("Policy %s modified successfully.  Fields updated were: %s", response.policyId, JSON.stringify(response.fieldsUpdated));
    } catch (err) {
        console.error("Failed to modify policy because: %s", JSON.stringify(err));
    }
}

const getPolicyExample = async () => {
    try {
        const response = await pmApiClient.getPolicy("user-account-access");
        console.log("Policy details are: %s", JSON.stringify(response));
    } catch (err) {
        console.error("Failed to query policy because: %s", JSON.stringify(err));
    }
}

const getManyPoliciesExample = async () => {
    try {
        const response = await pmApiClient.getManyPolicies() as PolicyListResponse;
        console.log("All known policies are: %s", JSON.stringify(response.result));
    } catch (err) {
        console.error("Failed to query policies because: %s", JSON.stringify(err));
    }
}

const getManyPolicyDefinitionsNoFilterExample = async () => {
    try {
        const response = await pmApiClient.getManyPolicies(undefined, undefined, undefined, true) as PolicyDataListResponse;
        // It is also possible to iterate through the policies and use their fields like response.result[0].agingOffsetUnit or response.result[0].policyId
        console.log("All known policy definitions are: %s", JSON.stringify(response.result));
    } catch (err) {
        console.error("Failed to query policy definitions because: %s", JSON.stringify(err));
    }
}

const getManyPolicyDefinitionsWithFilterExample = async () => {
    try {
        // Fetch 20 policy records where the policy ID starts "marketing-", skipping the first 10.  Include the policy definitions.
        const response = await pmApiClient.getManyPolicies("marketing-", 10, 20, true) as PolicyDataListResponse;
        // It is also possible to iterate through the policies and use their fields like response.result[0].agingOffsetUnit or response.result[0].policyId
        console.log("All known policy definitions are: %s", JSON.stringify(response.result));
    } catch (err) {
        console.error("Failed to query policy definitions because: %s", JSON.stringify(err));
    }
}

File: ruby-examples/policy_management_examples.rb
require 'metronome'

def create_policy_example(client)
    begin
        response = client.put_policy("user-account-access", {create_or_modify_policy_request: Metronome::CreateOrModifyPolicyRequest.new({aging_strategy: Metronome::AgingStrategy::LAST_READ_OR_WRITE, aging_offset_amount: 2, aging_offset_unit: Metronome::AgingOffsetUnit::YEAR})})
        puts "Created policy successfully: #{response}"
    rescue Metronome::ApiError => e
    puts "Exception when calling PolicyManagementApi->put_policy: #{e}"
    end
end

def modify_policy_example(client)
    begin
        response = client.put_policy("user-account-access", {create_or_modify_policy_request: Metronome::CreateOrModifyPolicyRequest.new({aging_strategy: Metronome::AgingStrategy::LAST_READ_OR_WRITE, aging_offset_amount: 3, aging_offset_unit: Metronome::AgingOffsetUnit::MONTH})})
        puts "Modified policy successfully: #{response}"
    rescue Metronome::ApiError => e
    puts "Exception when calling PolicyManagementApi->put_policy: #{e}"
    end
end

def get_policy_example(client)
    begin
        policy = client.get_policy("user-account-access")
        puts "Policy ID is #{policy.policy_id} created at #{policy.create_date} with aging strategy #{policy.aging_strategy} which ages by #{policy.aging_offset_amount} units of #{policy.aging_offset_unit}"
    rescue Metronome::ApiError => e
    puts "Exception when calling PolicyManagementApi->get_policy: #{e}"
    end
end

def get_many_policies_example(client)
    begin
        response = client.get_many_policies
        puts "There are #{response.result.length} policies which are #{response.result}"
    rescue Metronome::ApiError => e
    puts "Exception when calling PolicyManagementApi->get_many_policies: #{e}"
    end
end

def get_many_policy_definitions_no_filter_example(client)
    begin
        response = client.get_many_policies({include_data: true})
        puts "There are #{response.result.length} policies."
        response.result.each do |policy|
            puts "Policy ID is #{policy.policy_id} created at #{policy.create_date} with aging strategy #{policy.aging_strategy} which ages by #{policy.aging_offset_amount} units of #{policy.aging_offset_unit}"
        end
    rescue Metronome::ApiError => e
    puts "Exception when calling PolicyManagementApi->get_many_policies: #{e}"
    end
end

def get_many_policy_definitions_with_filter_example(client)
    begin
        response = client.get_many_policies({prefix: "marketing-", offset: 10, limit: 20, include_data: true})
        puts "There are #{response.result.length} policies."
        response.result.each do |policy|
            puts "Policy ID is #{policy.policy_id} created at #{policy.create_date} with aging strategy #{policy.aging_strategy} which ages by #{policy.aging_offset_amount} units of #{policy.aging_offset_unit}"
        end
    rescue Metronome::ApiError => e
    puts "Exception when calling PolicyManagementApi->get_many_policies: #{e}"
    end
end

# Configure the client
Metronome.configure do |config|
    config.api_key['ApiKeyAuth'] = ENV["METRONOME_API_KEY"]
    config.customer_id = "pds-internal-code-examples"
    config.region = "eu-north-1"
    config.use_production_instance = false      # Use your dev instance
end

pm_api_client = Metronome::PolicyManagementApi.new

get_many_policy_definitions_no_filter_example(pm_api_client)
